src/lib.rs

```diff
@@ -0,0 +1,389 @@
+///                         .build()
+///                         .unwrap();
+/// assert!(matcher.matches("a a").is_ok());
+/// assert!(matcher.matches("a b").is_err());
+/// ```
+pub struct FMBuilder<'a> {
+    ptn: &'a str,
+    options: FMOptions,
+}
+
+impl<'a> FMBuilder<'a> {
+    /// Create a new `FMBuilder` with default options.
+    pub fn new(ptn: &'a str) -> Result<Self, Box<dyn Error>> {
+        let options = FMOptions {
+            name_matcher: None,
+            ignore_leading_whitespace: true,
+            ignore_trailing_whitespace: true,
+        };
+        Ok(FMBuilder { ptn, options })
+    }
+
+    /// Add a name matcher `Some((ptn_re, text_re))` (or unset it with `None`). Defaults to `None`.
+    ///
+    /// Name matchers allow you to ensure that different parts of the text match without specifying
+    /// precisely what they match. For example, if you have output where you want to ensure that
+    /// two locations always match the same name, but the name is non-deterministic you can allow
+    /// the use of `$` wildcards in your pattern:
+    ///
+    /// ```rust
+    /// use {fm::FMBuilder, regex::Regex};
+    ///
+    /// let ptn_re = Regex::new("\\$.+?\\b").unwrap();
+    /// let text_re = Regex::new(".+?\\b").unwrap();
+    /// let matcher = FMBuilder::new("$1 b $1")
+    ///                         .unwrap()
+    ///                         .name_matcher(Some((ptn_re, text_re)))
+    ///                         .build()
+    ///                         .unwrap();
+    /// assert!(matcher.matches("a b a").is_ok());
+    /// assert!(matcher.matches("a b b").is_err());
+    /// ```
+    pub fn name_matcher(mut self, matcher: Option<(Regex, Regex)>) -> Self {
+        self.options.name_matcher = matcher;
+        self
+    }
+
+    #[doc(hidden)]
+    pub fn ignore_leading_whitespace(mut self, yes: bool) -> Self {
+        self.options.ignore_leading_whitespace = yes;
+        self
+    }
+
+    #[doc(hidden)]
+    pub fn ignore_trailing_whitespace(mut self, yes: bool) -> Self {
+        self.options.ignore_trailing_whitespace = yes;
+        self
+    }
+
+    /// Turn this `FMBuilder` into a `FMatcher`.
+    pub fn build(self) -> Result<FMatcher<'a>, Box<dyn Error>> {
+        Ok(FMatcher {
+            ptn: self.ptn,
+            options: self.options,
+        })
+    }
+}
+
+/// The fuzzy matcher.
+pub struct FMatcher<'a> {
+    ptn: &'a str,
+    options: FMOptions,
+}
+
+impl<'a> FMatcher<'a> {
+    /// A convenience method that automatically builds a pattern for you using `FMBuilder`'s
+    /// default options.
+    pub fn new(ptn: &'a str) -> Result<FMatcher, Box<dyn Error>> {
+        FMBuilder::new(ptn)?.build()
+    }
+
+    /// Does this fuzzy matcher match `text`? Returns `Ok(())` on success or `Err(usize)` on error
+    /// where the `usize` is the line number of the first line in `text` where the match failed.
+    pub fn matches(&self, text: &str) -> Result<(), usize> {
```

Comment:
Idea for future PR, use a custom error type whose "Display" formats the nature of the failing match nicely with some line context.